kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Reflector.java

```diff
@@ -59,54 +58,48 @@ private L getList() {
       .withTimeoutSeconds(null).build(), operationContext.getNamespace(), operationContext);
   }
 
-  public void listAndWatch() {
-    log.info("Started ReflectorRunnable watch for {}", apiTypeClass);
-    reListAndSync();
-    startWatcher();
-  }
-
   public void stop() {
     isActive.set(false);
+    stopWatcher();
+  }
+
+  private synchronized void stopWatcher() {
     Watch theWatch = watch.getAndSet(null);
     if (theWatch != null) {
+      String ns = operationContext.getNamespace();
+      log.debug("Stopping watcher for resource {} v{} in namespace {}", apiTypeClass, lastSyncResourceVersion.get(), ns);
       theWatch.close();
     }
   }
 
-  private void reListAndSync() {
+  /**
+   * <br>Starts the watch with a fresh store state.
+   * <br>Should be called only at start and when HttpGone is seen.
+   */
+  void listSyncAndWatch() {
     store.isPopulated(false);
     final L list = getList();
     final String latestResourceVersion = list.getMetadata().getResourceVersion();
+    lastSyncResourceVersion.set(list.getMetadata().getResourceVersion());
     log.debug("Listing items ({}) for resource {} v{}", list.getItems().size(), apiTypeClass, latestResourceVersion);
-    lastSyncResourceVersion.set(latestResourceVersion);
     store.replace(list.getItems(), latestResourceVersion);
+    startWatcher(latestResourceVersion);
   }
 
-  private void startWatcher() {
-    log.debug("Starting watcher for resource {} v{}", apiTypeClass, lastSyncResourceVersion.get());
-    Watch theWatch = watch.getAndSet(null);
-    if (theWatch != null) {
-      theWatch.close();
-      log.debug("Stopping previous watcher, and delaying execution of new watcher");
-      try {
-        Thread.sleep(WATCH_RESTART_DELAY_MILLIS);
-      } catch (InterruptedException e) {
-        log.debug("Reflector thread was interrupted: {}", e.getMessage());
-        Thread.currentThread().interrupt();
+  private synchronized void startWatcher(final String latestResourceVersion) {
+    if (!isActive.get()) {
```

Comment:
nit: this variable is only used to mark that the Reflector is stopped. Would it make sense to rename it to `isStopped`?

Reply:
I've updated it to be stopped.  running would also make sense.

Reply:
It was really a nit. Not that the previous name didn't make sense, but in the context of the code review and within the GH interface I had absolutely no idea what that flag was supposed to be doing (once I opened the full file it made more sense).

Reply:
In fact, after seeing that the variable is also checked in the last method `isRunning`, `running` does make more sense indeed.